python/smqtk/algorithms/nn_index/mrpt.py

```diff
@@ -0,0 +1,451 @@
+        self._descr_cache_filepath = normpath(descriptor_cache_filepath)
+        # Now they're either None or an absolute path
+
+        # parameters for building an index
+        self._depth = depth
+        self._num_trees = num_trees
+        self._required_votes = required_votes
+
+        self._trees = []
+
+        # In-order cache of descriptors we're indexing over.
+        # - index will spit out indices to list
+        #: :type: list[smqtk.representation.DescriptorElement] | None
+
+        #: :type: None | int
+        self._rand_seed = None
+        if random_seed:
+            self._rand_seed = int(random_seed)
+
+        # The process ID that the currently set MRPT instance was built/loaded
+        # on. If this differs from the current process ID, the index should be
+        # reloaded from cache.
+        self._pid = None
+
+        # Load the index/parameters if one exists
+        if self._has_model_files():
+            self._log.info("Found existing model files. Loading.")
+            self._load_mrpt_model()
+
+        if (self._descr_cache_filepath and
+                osp.isfile(self._descr_cache_filepath)):
+            self.load_descr_cache(self._descr_cache_filepath)
+
+    def get_config(self):
+        return {
+            "index_filepath": self._index_filepath,
+            "parameters_filepath": self._index_param_filepath,
+            "descriptor_cache_filepath": self._descr_cache_filepath,
+            "random_seed": self._rand_seed,
+            "pickle_protocol": self._pickle_protocol,
+            "depth": self._depth,
+            "num_trees": self._num_trees,
+            "required_votes": self._required_votes,
+        }
+
+    def _has_model_files(self):
+        """
+        check if configured model files are configured and exist
+        """
+        return (self._index_filepath and osp.isfile(self._index_filepath) and
+                self._index_param_filepath and osp.isfile(self._index_param_filepath))
+
+    def build_index(self, descriptors):
+        """
+        Build the index over the descriptor data elements.
+
+        Subsequent calls to this method should rebuild the index, not add to
+        it, or raise an exception to as to protect the current index.
+
+        :raises ValueError: No data available in the given iterable.
+
+        :param descriptors: Iterable of descriptor elements to build index
+            over.
+        :type descriptors:
+            collections.Iterable[smqtk.representation.DescriptorElement]
+
+        """
+        super(MRPTNearestNeighborsIndex, self).build_index(descriptors)
+
+        self._log.info("Building new MRPT index")
+
+        self._log.debug("Storing descriptors")
+        self.init_descr_cache(descriptors, self._descr_cache_filepath,
+                              pickle_protocol=self._pickle_protocol)
+
+        self._log.debug("Accumulating descriptor vectors into matrix")
+        # XXX is an interval of 1.0 really a good idea?
```

Comment:
Ah, I misunderstood the interpretation of the parameter. I thought it was "report every element."